blockchain/chain_sync/src/sync.rs

```diff
@@ -424,10 +424,10 @@ where
             .await;
 
         // Only update target on initial sync
-        if self.state == ChainSyncState::Bootstrap {
+        if *self.state.lock().await == ChainSyncState::Bootstrap {
             if let Some(best_target) = self.select_sync_target().await {
                 self.schedule_tipset(best_target).await;
-                self.state = ChainSyncState::Initial;
+                *self.state.lock().await = ChainSyncState::Initial;
```

Comment:
I believe this keeps the lock and will deadlock here, has this been tested?

Reply:
Not tested, but I'm quite certain that `if *self.state.lock().await` does not keep the guard around since it's not bound to a variable, if that's what you're referring to. ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f082ff7b18953ba225181e2d96df9f9) that uses std's Mutex)

Reply:
I've had issues with async-std mutexes holding the lock in this exact case in the past, not sure if it's changed or only applies to `RwLock` but can double check

Reply:
nvm, there was a specific case where it deadlocked similar to this, and moving the read outside of the `if` removed the deadlock, I just tested and this works

Reply:
Oh yeah, I can totally see that happening. Thanks for pointing it out either way, I could have easily messed this up